### ISSUE TITLE
Multiple Selection

### ISSUE THREAD

**Issue by pubpub-zz** [pubpub-zz/ppInk#97]
> Selected objects:
> One of my favorite options in ppink is when I select, move and copy strokes, figures or something that I drew on the screen. That's fantastic! But sometimes I only needed to select two or three objects at a time to move them to another place on the screen. I don't know how difficult it would be implement a selector in orden to increase this marvelous function. I would be fantastic if you could implement it. For now we can move an object, copy an object and move all the objects that are on the screen.

I agree this could be interesting, but this may be quite complex for both implementation and Operator interaction(I want to keep the interface compatible with a full tablet(touch only) mode) Will think about it. 👍 

_Originally posted by @pubpub-zz in https://github.com/pubpub-zz/ppInk/issues/95#issuecomment-844534760_

> -Regarding selection of strokes: When i make a drawing It consist of couple of strokes. Lets say I want to move this drawing I can use hotkey 'P'. After that i make another drawing with couple of strokes, now if i want to move it, it moves everything on screen including the first drawing that i have made before. This behavior is obvious though. But it can be avoided if there is facility to select multiple strokes at a time. Right now only one stroke can be selected at a time which is not really helpful, because drawing is made of more than one stroke. So if possible 'drag and select' function is required which is similar to when we select multiple folders on desktop.

_Originally posted by @pubpub-zz in https://github.com/pubpub-zz/ppInk/issues/92#issuecomment-846431185_

**Comment by pubpub-zz** [comment]
@chrisj0192 @gitsamhub

I've found a solution to implement  multi Selection (first alpha version partially coded):
[ppInk64_MultiSel06a.zip](https://github.com/pubpub-zz/ppInk/files/6604980/ppInk64_MultiSel06a.zip)

In order to use it : there is a new icon (i've reduced moving Icon to not take to much place)
Currently there is only the lasso (I intend to add also the hovering to add one stroke at a time).
You add to selection with left click, and right click to remove
at the end of the selection, you can click on the eraser to delete the selection, else move or copy allow to move object

You can click on undo to cancel the last selection modification.
For the moment there is no shortcut

Can you give your feed back on usability

**Comment by NOVAinc** [comment]
I've tested this new feature and it works as intended. I would suggest being able to select these one at a time but I see you have that on the roadmap already. Your suggestion of hitting the eraser to move selected strokes seems pretty good to me!

I'd personally like if you could see the combined length of the strokes you have selected, but it seems very much usable from what I've seen!

**Comment by chrisj0192** [comment]
Hello @pubpub-zz I'm testing your beta version in different cases, I'll post my observations about the multiple selector, pen width, load/save shortcuts and snapshot in a few hours.

**Comment by gitsamhub** [comment]
Hi @pubpub-zz 
The Lasso function is working as expected. I see you are already working towards adding hotkey for it. Right click deselecting idea also seems intuitive.

Now this is unrelated to the current topic but I would like to mention it here.
The issue is regarding 'Hands-free Drawing'.

How it is supposed to work: Lets say I press '1' to select pen and draw few strokes on screen. After that I press 'H' to activate hands-free drawing. I draw something with it. Now if I press 1 again , It doesn't change to pen. If I press any number key then it should activate that particular pen. 

I think cycling through Hands free drawing ( Stranded, black and white) is not required, as user can choose desired color with color picker or pen tool hotkeys. Ideally, if i press any number key(1,2,3 etc.) in any point of time it should switch to pen tool and not just its color.

I also noticed Hands free drawing produce jagged edges to the drawing. I guess it is an anti-aliasing issue.

Thank you.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz, regarding the lasso function. I like how it works. 

>Currently there is only the lasso (I intend to add also the hovering to add one stroke at a time).

I'd be pretty nice!I I'll wait for it and also I'll wait for a hotkey to test it much better in order to get more details on usability.

It's really intuitive to add an element/stroke to selection with left click, and use right click to remove, and the idea of erasing selected strokes by selecting eraser is nice too.

I was testing the lasso by combining it with the move/copy functions and it works as expected.

As I told you above I'll wait for a hotkey option to do more tests and get more results.

For now, it works as expected!

Thanks!

**Comment by pubpub-zz** [comment]
> How it is supposed to work: Lets say I press '1' to select pen and draw few strokes on screen. After that I press 'H' to activate hands-free drawing. I draw something with it. Now if I press 1 again , It doesn't change to pen. If I press any number key then it should activate that particular pen.
> 
> I think cycling through Hands free drawing ( Stranded, black and white) is not required, as user can choose desired color with color picker or pen tool hotkeys. Ideally, if i press any number key(1,2,3 etc.) in any point of time it should switch to pen tool and not just its color.

Pens buttons selects Color/Transparency/Width/Linestyle /fading/... without changing tools  which are independents, else how can you change color without changing tool. Consider thoses as Pen button and not Pen tool.

Color picker is designed to modify the current pen definition not the standard way to select color It is a quicker way proposed to accelerate color selection which is  normally done by the color picker in the pen dialogbox.

> I also noticed Hands free drawing produce jagged edges to the drawing. I guess it is an anti-aliasing issue.

Yes... Not sure the problem is not part of Ink Library...

**Comment by pubpub-zz** [comment]
> I'd personally like if you could see the combined length of the strokes you have selected, but it seems very much usable from what I've seen!

Part of the coding in progress 😊

**Comment by pubpub-zz** [comment]
New alpha version with hotkeys and measurement
[ppInk64_MultiSel07a.zip](https://github.com/pubpub-zz/ppInk/files/6612286/ppInk64_MultiSel07a.zip)

**Comment by NOVAinc** [comment]
This is already perfect IMO. The only thing I noticed was that lasso selection doesn't add the length, only click select. Great work :)

**Comment by pubpub-zz** [comment]
have you seen 2nd line "Total" when hovering?

**Comment by NOVAinc** [comment]
Hmmm. I tested again and now it seems to be working. Upon first trying it, the second line wouldn't appear when strokes were selected using lasso. If I can reproduce I'll let you know. For now it seems to be working as intended, then. Lasso can get a bit laggy when you're drawing it though. Have you noticed this? If it's just me I can make a quick video.

**Comment by pubpub-zz** [comment]
yes pls

**Comment by NOVAinc** [comment]
I think lag should be visible in this clip
https://user-images.githubusercontent.com/19545604/121101716-c1001600-c7d2-11eb-8a5f-4ca933de5f3c.mp4

**Comment by chrisj0192** [comment]
Hello @pubpub-zz  it seems the right click to remove elements don't work. I'm doing many tests. Later I'll show you my results.

**Comment by chrisj0192** [comment]
I was testing the usability of the lasso with hotkeys and combinations  in many cases (around 30+) and It's incredible. I have no words to express this amazing feature. My congrats for it.

I have some observations related to the lasso an others (not  related to the lasso)
I'd like to show you some bugs  (I'm not sure if they are bugs) but I'll show them to you in this video:
https://user-images.githubusercontent.com/20199672/121160046-60dc9480-c811-11eb-981d-296a69818e25.mp4

My intention when I make these test videos is you to improve some bugs or defects that I find (possible bugs, I don't know, If possible of course) because, to be honest your software is the best of all programs (including online educational whiteboards) I've found on the internet so far. Honestly using ppInk, a class is far more dynamic than using online whiteboards and pdf annotators. Well that's my point of view.

Thanks for your attention.
Have a nice day.

PS: I have an observation about the clipArt too but I will comment it more later.

**Comment by chrisj0192** [comment]
As @NOVAinc already told you. Today, another issue I've notived is about lagging when using the lasso with a regular amount of strokes by selecting and combining move/copy options. I confirm that it happens to me too. Even so it works excellent.

**Comment by pubpub-zz** [comment]
> I have some observations related to the lasso and others (not related to the lasso)
> I'd like to show you some bugs (I'm not sure if they are bugs) but I'll show them to you in this video:
> https://user-images.githubusercontent.com/20199672/121160046-60dc9480-c811-11eb-981d-296a69818e25.mp4

Odd. I have to try to reproduce...

**Comment by pubpub-zz** [comment]
New beta fixing Text Move in groups and also should also fix the issue with pend dialog box opening on clipart reported @chrisj0192 

[ppInk64_MultiSel09a.zip](https://github.com/pubpub-zz/ppInk/files/6623394/ppInk64_MultiSel09a.zip)

**Comment by chrisj0192** [comment]
@pubpub-zz I've been testing ppInk64_MultiSel09b these last days and I got these results:
- As I told you in #106 , Pen dialog box opening on clipart shortcut was fixed. I've not had this problem again.
- Regarding text/letters move in groups using the lasso, it's working well sometimes. Although other times I noticed when selecting a text/letters using the lasso and then moving them with my custom shortcut, they were not displaced (this problem happens sometimes).
Here you are a short videos about it.
Test 1:
https://user-images.githubusercontent.com/20199672/121762430-74cb1380-cafb-11eb-846e-ffadbdad6547.mp4

Test 2:
https://user-images.githubusercontent.com/20199672/121762433-785e9a80-cafb-11eb-8e35-b8e55db0a4a2.mp4

odd.

**Comment by pubpub-zz** [comment]
@chrisj0192,
I think I've found your issue but not able to reproduce it. can you try it ?
[ppInk64_varFix14a.zip](https://github.com/pubpub-zz/ppInk/files/6651439/ppInk64_varFix14a.zip)
thanks

**Comment by pubpub-zz** [comment]
@chrisj0192, any update?

**Comment by chrisj0192** [comment]
Hello @pubpub-zz long time no see, then I'll upload a bunch of massive tests I've been doing these last few days. Sorry for not answering you, but my current job makes me feel exhausted. 😫
btw, Has you changed anything in the hotkeys option (ppInk64_varFix14a.zip)? because when pressing the backspace key, it is taken as a shortcut, in previous versions that didn't happen.

**Comment by pubpub-zz** [comment]
Yes, I wanted to be able to make hotkeys with escape, del or backspace. to delete hotkey now it is ctrl+shift+del or  ctrl+shift+backspace
 I forgot to update languages...

**Comment by chrisj0192** [comment]
Hello @pubpub-zz
> Yes, I wanted to be able to make hotkeys with escape, del or backspace. to delete hotkey now it is ctrl+shift+del or ctrl+shift+backspace
I forgot to update languages...

I see, ok, thanks.

Well, I was testing  and I get some strange results about multiple selection (one by one). I've noticed when there are many strokes or elements the selection is a little bit more difficult.

Test 1:
I don't know if this video is clear showing about it.
https://user-images.githubusercontent.com/20199672/123538116-caf1a680-d6f8-11eb-98a3-33f63b2a5e1f.mp4

Test 2:
When selecting strokes using multiple selection It seems that sometimes the lasso doesn't select elements and I've noticed it happens when the lasso gets tangled or matted? (I'm not sure which the correct word is) this is an example.
image:
![image](https://user-images.githubusercontent.com/20199672/123538357-2a03eb00-d6fa-11eb-96c9-1fba42ada016.png)
video:
https://user-images.githubusercontent.com/20199672/123538208-510ded00-d6f9-11eb-942f-d071acea5ec6.mp4

Test 3:
Another test about tangled selection.
https://user-images.githubusercontent.com/20199672/123538310-e3ae8c00-d6f9-11eb-9bee-490aa1d4cd89.mp4

Test 4 and test 5:
I was trying combine one selection with multiple selection in these tests:
https://user-images.githubusercontent.com/20199672/123538406-66374b80-d6fa-11eb-89e5-8b3b4af679db.mp4

https://user-images.githubusercontent.com/20199672/123538442-91ba3600-d6fa-11eb-8295-fd006318dfab.mp4

Test 6:
Another problem that I've noticed is about lagging when selecting a lot of strokes.
Here you are the video:
https://user-images.githubusercontent.com/20199672/123538484-bdd5b700-d6fa-11eb-8660-361e9d8933a1.mp4

In summary:
- As I told you about matted or tangled selection, it happens frequently when there are a lot of strokes on the screen or well, I suspect that.
- Another problem is that lag is frequent when there are many strokes, I showed as an example a geometry exercise because I usually do exercises with many strokes and lag is frequent in this case.
- It seems the group selection was fixed. I've not had any problems about it so far.

**Comment by pubpub-zz** [comment]
I'm working on some performances improvments : can you tell me if it is better with this version ? 
[ppInk64_perf05b.zip](https://github.com/pubpub-zz/ppInk/files/6766333/ppInk64_perf05b.zip)

**Comment by pubpub-zz** [comment]
@chrisj0192 
new version.
I've added more checks to make mouseDown,and other events: From My tests the  hickups have reduced. Can you tell me with your computer if it is better?

**Comment by chrisj0192** [comment]
Hello @pubpub-zz 
I was testing ppInk64_apng_perfo06b.zip and yes I confirm that the hickups issues were greatly reduced. it's definitely much better. it works fluider too.

Btw, I was wondering if it's feasible to add some math functions in edit text or a math text editor? (basic functions only)

Thank you.

**Comment by pubpub-zz** [comment]
implementing a math editor is too specific, even math functions.
What I would propose would be :
* engage pointer mode
* use a math editor such as libreoffice to build an image
* export the formula as an image
* disengage pointer mode
* paste as image

in order to ease the process I can add an option to add the clipboard to the clipart, select it and be read to draw it when returning from pointer mode.
Your opinion ?

**Comment by xxBloodyMaryxx** [comment]
> @chrisj0192
> new version.
> I've added more checks to make mouseDown,and other events: From My tests the hickups have reduced. Can you tell me with your computer if it is better?
> [ppInk64_apng_perfo06b.zip](https://github.com/pubpub-zz/ppInk/files/6772517/ppInk64_apng_perfo06b.zip)

Hey thanks for this, I has just started using this and this feels already so great. I do have some suggests and comments.

Clipart - 
- The clipart images doesn't move when I use the pan 3 mode (moving the entire screen). Apparently, the image (what is displayed) doesn't move but the referenced location did move because when I tried to use lasso, I couldn't select the visible image, but I was able to select an invisible slot that corresponds to the image.
- This is more of a suggestion: Would it be possible to add image in Clipboard directly onto screen, which having to go to ClipArt. I would love to use Windows Snipping Tool followed by Ctrl+V to add the screensnips. Also along the same line, it would be interesting if these can be placed by respecting the Ratio Aspect of the image. Forget it if that's very complicated.

Zoom Tool - 
- The Zoom 2 mode, where you draw a rectangle and the screen zoom in on it, doesn't seem to zoom into the region I marked. There seems to be some offset towards top-left corner.
- Also, if I write something in the zoom 2 mode and then zoom back out using the hotkey, all the writings seems to disappear. Is that intentional?

Edit Tool - 
- Since you already have a well implemented Lasso Tool, would it be possible to couple that with Edit Tool as well. As in say I make multiple copies of a particular drawing and I want to change the color of each copy (say to represent various timeframes or something), then this feature could be pretty handy I guess.

Bug in Docking/Undocking Toolbar- 
- I started ppInk using the Global Hotkey (Mouse pointer Toggling), docked the toolbar using the hotkey, now if I want to toggle back to the mouse point, if the use the Global Hotkey it seems to undock the toolbar rather than toggling the mouse pointer. 
- And so everything I use the Global Hotkey to move into ppInk, it undocks the Toolbar. 

I will let you know if I find anything more. Overall this is so cool, its way more than what I thought would be possible.

**Comment by pubpub-zz** [comment]
> Hey thanks for this, I has just started using this and this feels already so great. I do have some suggests and comments.

> Clipart -
> 
>     * The clipart images doesn't move when I use the pan 3 mode (moving the entire screen). Apparently, the image (what is displayed) doesn't move but the referenced location did move because when I tried to use lasso, I couldn't select the visible image, but I was able to select an invisible slot that corresponds to the image.

Reproduced. I will try to fix it very next release

>     * This is more of a suggestion: Would it be possible to add image in Clipboard directly onto screen, which having to go to ClipArt. I would love to use Windows Snipping Tool followed by Ctrl+V to add the screensnips. Also along the same line, it would be interesting if these can be placed by respecting the Ratio Aspect of the image. Forget it if that's very complicated.

Quite in line with my proposal https://github.com/pubpub-zz/ppInk/issues/97#issuecomment-876775178
However this will probably not be in very next release. 
PS : Windows Snipping Tool sucks! ppInk's is better 😆 

> Zoom Tool -
> 
>     * The Zoom 2 mode, where you draw a rectangle and the screen zoom in on it, doesn't seem to zoom into the region I marked. There seems to be some offset towards top-left corner.

can you provide video ?

> 
>     * Also, if I write something in the zoom 2 mode and then zoom back out using the hotkey, all the writings seems to disappear. Is that intentional?

when zooming, you capture in zoom2, you snapshot the area, expand it and allow drawing over, but your drawing can't be restored like that for the moment...

> 
> Edit Tool -
> 
>     * Since you already have a well implemented Lasso Tool, would it be possible to couple that with Edit Tool as well. As in say I make multiple copies of a particular drawing and I want to change the color of each copy (say to represent various timeframes or something), then this feature could be pretty handy I guess.

this is an idea, not for next release. Can you open a new issue for tracking


> 
> Bug in Docking/Undocking Toolbar-
> 
>     * I started ppInk using the Global Hotkey (Mouse pointer Toggling), docked the toolbar using the hotkey, now if I want to toggle back to the mouse point, if the use the Global Hotkey it seems to undock the toolbar rather than toggling the mouse pointer.

this is because "engage Pointer and auto-fold with Alt+Tab" is set in options.

> 
>     * And so everything I use the Global Hotkey to move into ppInk, it undocks the Toolbar.

this is because "engage Pointer and auto-fold with Alt+Tab" is set in options.

Globally expected for same reason
 
> I will let you know if I find anything more. Overall this is so cool, its way more than what I thought would be possible.

Happy you enjoy 😊

**Comment by xxBloodyMaryxx** [comment]
Hey, 

Debugging background - 
- Win 10 ver 20H2
- OBS and keylogging plugin
- MS Paint
- Settings changed in ppink options: Global Mouse Toggle Hotkey changed to Ctrl+Q, and some other insignificant icon removals in Toolbar. Engage point and auto-fold is disabled.
- I can't access my pen tablet right now, and had to use laptop touch pad, so ignore the many unanswered mouse clicks.

Here is a video of the zoom bug that I was able to reproduce:

https://user-images.githubusercontent.com/72195985/125112691-22364480-e105-11eb-816b-2f707a675bd1.mp4

There seems to be something wrong with the zoom 1 mode as well, I don't remember that the previous time I checked it out, but now it seems to persist even after exiting and restarting many times, probably I might have missed it the first time. I have shown the options before and after (both are same) just to be complete.

Here is a recording of the docking undocking bug that I had mentioned about :

https://user-images.githubusercontent.com/72195985/125113067-c3bd9600-e105-11eb-8ec3-f2d5586591c5.mp4

Notice that the every first time after I toggled to Mouse pointer, my very first mouse click was answered by by toggling back into ppink overlay, although this is pretty infrequent. I saw this the during my first time as well, but was unable to reproduce it so didnt report it. And the second part which is undocking the toolbar when its supposed to toggle to mouse pointer mode. I have shown the options before and after (both are same) just to be complete. 

Do you want me to try this with Engage point and auto-fold enabled? I didn't understand this feature so had it turned of the first time. 

Hope this is helpful. Let me know if I should do anything more. 

> Can you open a new issue for tracking

Sure I'll do that now.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz 

> implementing a math editor is too specific, even math functions.
What I would propose would be : (...)

I'd done those actions before hehe and yes I like the idea, but that's why I was wondering if it's possible to add a math editor. Actually, it's not necessary because I use handfree drawing to make strokes related to math operators (it's more comfortable for me, although it would be more comfortable if I had a graphics tablet soon hehe) but I agree with you to add an option to add the clipboard to the clipart, select it and be read to draw it when returning from pointer mode (I like this idea).
I just realized that I can put a image to the clipArt from clipboard (image taken using snapshot with white background) I like this function because an imagen loaded from the clipboard appears without background but this image is not saved to files, I mean it appears in clipArt but not as a file and I think that's the reason why sometimes I get a dialog window related to no image upload or something like that when I press clipArt button in options. It would be nice to save images loaded from clipboard to files.

Another thing that I found very necessary is a sort option, because I have more than 100 images in my clipArt hehe. It would be nice to have an option to sort them by name (in my case). I use codes to identify the names of my images.

On the other hand, I found a problem related to fading and eraser. When I use fading and immediately swap to the eraser to delete some strokes (using alt), this dialog window shows up.
![eraser-fading](https://user-images.githubusercontent.com/20199672/125129456-1dea3700-e0c5-11eb-94ad-6260a1f10706.png)

As xxBloodyMaryxx told you, Another one is when I use pan function (move all strokes), Images loaded from the clipArt don't move (I forgot to report this one).

Finally, a question, is it possible to have a custom color palette? (It would be great to quickly change from one color to another) because we only have 10 pens and I use other colors for the highlighters.

Thanks in advance.

**Comment by pubpub-zz** [comment]
@chrisj0192 

> I just realized that I can put a image to the clipArt from clipboard (image taken using snapshot with white background) I like this function because an imagen loaded from the clipboard appears without background but this image is not saved to files, I mean it appears in clipArt but not as a file and I think that's the reason why sometimes I get a dialog window related to no image upload or something like that when I press clipArt button in options. It would be nice to save images loaded from clipboard to files.

Can you create a new issue for this please. may come in 1.8, not 1.7



> Another thing that I found very necessary is a sort option, because I have more than 100 images in my clipArt hehe. It would be nice to have an option to sort them by name (in my case). I use codes to identify the names of my images.

There is two things:
a) you can type the beggining of the file and you will jump to them
b) the order in the list is the order of the ini file . you can organize them in a clever manner to find them easily
c) you can modify in the config.ini file this parameter:
                       ImageStamp_Size = 128
if you set it to a lower value the thumbnails in the list will be smaller which may help you to find your image.

I tried it in

> On the other hand, I found a problem related to fading and eraser. When I use fading and immediately swap to the eraser to delete some strokes (using alt), this dialog window shows up.

Problem reproduced and fixed(normally)

> As xxBloodyMaryxx told you, Another one is when I use pan function (move all strokes), Images loaded from the clipArt don't move (I forgot to report this one).

as promised, it is fixed

> Finally, a question, is it possible to have a custom color palette? similar to the ms paint palette (It would be great to quickly change from one color to another) because we only have 10 pens and I use other colors for the highlighters.

My recommandation would be to make a bat program using curl or a powershell script  that would use the Rest API to modify the pens as required. Then you would just have to run it (even in pointer mode) to modify pens definitions.

[ppInk64_varFixes09b.zip](https://github.com/pubpub-zz/ppInk/files/6794020/ppInk64_varFixes09b.zip)

**Comment by pubpub-zz** [comment]
@xxBloodyMaryxx,
as said about the issue of the pan with images is fixed
[ppInk64_varFixes09b.zip](https://github.com/pubpub-zz/ppInk/files/6794020/ppInk64_varFixes09b.zip)

 about you zoom issue can you tell how many screens you are using and what is the scale you are using



> Notice that the every first time after I toggled to Mouse pointer, my very first mouse click was answered by by toggling back into ppink overlay, although this is pretty infrequent. I saw this the during my first time as well, but was unable to reproduce it so didnt report it. And the second part which is undocking the toolbar when its supposed to toggle to mouse pointer mode. I have shown the options before and after (both are same) just to be complete.

apparently ppInk does not received focus. when required. I've added extra code, can you tell me if it is better

**Comment by xxBloodyMaryxx** [comment]
ppInk64_varFixes09b.zip review - 

Pan Bug in Clipart - 
Now the clipart does move, but seems to cluster together at one end, here is the video

https://user-images.githubusercontent.com/72195985/125158014-cbc01900-e18b-11eb-81e2-d9d2305fcfa4.mp4

I am using 1080p default laptop screen. I haven't configured any other display in this machine. I have reproduced it with ppInk64_varFixes09b.zip: 

https://user-images.githubusercontent.com/72195985/125157841-d62de300-e18a-11eb-8697-fd8609633bdb.mp4

Yes, I think after toggling to Mouse pointer (which apparently only after first getting the toolbar undocked), it seems to require a click to move the focus out. Following is what I could capture - 

https://user-images.githubusercontent.com/72195985/125157919-60764700-e18b-11eb-8230-bca30077ec45.mp4

The Pointer and Autofold was disabled for this last recording.
I am guessing this issue might be with the automatic docking/undocking feature.

**Comment by pubpub-zz** [comment]
about the issue with the zoom, I reminds me of #92, colorpicker issue. in a word the issue fix should be change the compatibility / advanced PPP compatibility in ppInk.exe properties
![image](https://user-images.githubusercontent.com/4083478/125160319-8f38f100-e17c-11eb-9832-5f664b8df596.png)

About the click, after pointer mode, can you check that another program (maybe your key dispay) is not interfering

Finaly about pan and typo, tripple oups!!!!!!!!! should be good now
[ppInk64_pan10a.zip](https://github.com/pubpub-zz/ppInk/files/6795084/ppInk64_pan10a.zip)

**Comment by YouFoundJK** [comment]
Reviewing ppInk64_edit10b.zip - 

Amazing, the advanced compatibility option did fix the issue with both zoom modes. Pan is also working great now.

The only bug that's left is the docking/undocking one I think.

I noticed a pattern:

https://user-images.githubusercontent.com/72195985/125172408-30ed2c00-e1d7-11eb-9cfd-0512ea694253.mp4

During every first attempt after starting ppink.exe afresh, the global hotkey toggles the mouse pointer WITHOUT undocking the toolbar, the next mouse click does bring back focus to other windows (verified by spacebars that scroll the webpage) but at the same time activates the ppink again (two actions simultaneously). This happens only for first attempts.

For furthur attempts, the global hotkey first undocks the toolbar, and subsequently toggles the mouse pointer while the toolbar is undocked. There is a click going unanswered where the focus is getting shifted to the ppink overlay from ppink toolbar.

I am positive that there are no applications (atleast those I am aware of) that are interfering including the key logging since the was able to reproduce the same when the keylogging wasn't running. Also, I tried doing this with the toolbar disabled (within the config,ini) and everything was as expected, no issues there.

**Comment by chrisj0192** [comment]
Hello @pubpub-zz 
> Can you create a new issue for this please. may come in 1.8, not 1.7

Ok, I'll create it.

> There is two things:
a) you can type the beggining of the file and you will jump to them
b) the order in the list is the order of the ini file . you can organize them in a clever manner to find them easily
c) you can modify in the config.ini file this parameter:
ImageStamp_Size = 128
if you set it to a lower value the thumbnails in the list will be smaller which may help you to find your image.

Thanks for your recommendation, I've already tried it 😊

> Problem reproduced and fixed(normally)

I confirm the eraser vs fading issue is fixed now. it works perfectly swapping them.

> as promised, it is fixed

About pan function, I saw this issue was reported and now you've fixed it, In ppInk64_pan10a.zip, I confirm Pan function works perfectly!

> My recommandation would be to make a bat program using curl or a powershell script that would use the Rest API to modify the pens as required. Then you would just have to run it (even in pointer mode) to modify pens definitions.

I'm going to try your recommendation. Thank you 😊

Finally I detected a little bit problem with the color picker (it's the first time I use it and with the alt key the color picker gets stuck and without the alt key it also gets stuck. I need to press twice to get back to the canvas cursor. I'm not sure  if it's a normal function, could you try it?

Thank you for your amazing work!

**Comment by chrisj0192** [comment]
Hello @pubpub-zz 
> Can you create a new issue for this please. may come in 1.8, not 1.7

Ok, I'll create it.

> There is two things:
a) you can type the beggining of the file and you will jump to them
b) the order in the list is the order of the ini file . you can organize them in a clever manner to find them easily
c) you can modify in the config.ini file this parameter:
ImageStamp_Size = 128
if you set it to a lower value the thumbnails in the list will be smaller which may help you to find your image.

Thanks for your recommendation, I've already tried it 😊

> Problem reproduced and fixed(normally)

I confirm the eraser vs fading issue is fixed now. it works perfectly swapping them.

> as promised, it is fixed

About pan function, I saw this issue was reported and now you've fixed it, In ppInk64_pan10a.zip, I confirm pan function works perfectly!

> My recommandation would be to make a bat program using curl or a powershell script that would use the Rest API to modify the pens as required. Then you would just have to run it (even in pointer mode) to modify pens definitions.

I'm going to try your recommendation. Thank you 😊

Finally I detected a little bit problem with the color picker, it's the first time I use it and with the alt key the color picker gets stuck and without the alt key it also gets stuck. I need to press twice to get back to the canvas cursor. I'm not sure  if it's a normal function, could you try it?

Thank you for your amazing work!

**Comment by pubpub-zz** [comment]
> I am positive that there are no applications (atleast those I am aware of) that are interfering including the key logging since the was able to reproduce the same when the keylogging wasn't running. Also, I tried doing this with the toolbar disabled (within the config,ini) and everything was as expected, no issues there.

Have you try to unload this program:
![image](https://user-images.githubusercontent.com/4083478/125174468-0d6db580-e1c6-11eb-913d-66e994b99239.png)

**Comment by YouFoundJK** [comment]
Unload as in? I am using this:

https://github.com/xxdocobxx/DisplayKeystroke

Its a portable executable file that goes off grid as soon as I close (atleast it goes off task manager). Is this the one causing the issue?

**Comment by pubpub-zz** [comment]
I meant quit the program
I did some test with and without, and I've got reaction at first click.

**Comment by pubpub-zz** [comment]
I've prepared a test program to show active window.
To use it go to option/general, press on debug window to open 
in the window every sec, the active window title will be displayed. If ppink is active and has focus, it will display form1
When you will switch pointer on / off wait a little to display the current window
[ppInk64_invst11a.zip](https://github.com/pubpub-zz/ppInk/files/6796608/ppInk64_invst11a.zip)

**Comment by YouFoundJK** [comment]
Oh okay interesting. This is what I have got with ppInk64_invst11a.zip

https://user-images.githubusercontent.com/72195985/125190917-47d66180-e25d-11eb-9135-264b6419c656.mp4

Is it an issue with my Global Hotkey maybe?

**Comment by pubpub-zz** [comment]
Hi @YouFoundJK,
I've reproduced the problem but the fix is tricky. will come back ASAP

**Comment by pubpub-zz** [comment]
@YouFoundJK, @NOVAinc ,

Can you test this version : 
[ppInk64_varFix14a.zip](https://github.com/pubpub-zz/ppInk/files/6815646/ppInk64_varFix14a.zip)

I did some change on the pointermode. The first test case raised by @YouFoundJK is when you enter/exit pointer mode with the toolbardock. The issues was some requirement to click many times to get pen active. Note that is normal that the toolbar stays folded if it was folded before entering pointer mode.
Give me your feedbacks.

**Comment by NOVAinc** [comment]
I have tested it and it seems pretty good! Performance is great and the issues I noticed I haven't run into with this one so I think they might be gone. I'll make a switch to this version for work so I'll get more testing done!

**Comment by pubpub-zz** [comment]
Looking forward to see some video with animated gif

**Comment by pubpub-zz** [comment]
Waiting for a few feed back and one translation and I will issue 1.7RC2

**Comment by YouFoundJK** [comment]
Insane, the docking thing is fixed, and I couldn't find any more issue. This seems to be pretty stable.

@pubpub-zz How come only I had that changing focus issue? Is it some Windows setting that I might have messed up with?

**Comment by pubpub-zz** [comment]
@YouFoundJK,
Good!
The focus get/loose is quite tricky because it implicates some windows events. All my tests were based on engaging pointer mode from the toolbar from a undocked state. It is when I did some test from there I've been able to reproduced.With other people complaining about It was fixed replacing some key capture programs aside.
fixed in 68fda959f483d20dce79c9a1f269b323e4bfbad1

**Comment by chrisj0192** [comment]
Hello @pubpub-zz  I hope you are well, it seems you didn't read my last comment, I'd like to know your answer.
it's this one:
https://github.com/pubpub-zz/ppInk/issues/97#issuecomment-877692020
especifically, this part:
> Finally I detected a little bit problem with the color picker, it's the first time I use it and with the alt key the color picker gets stuck and without the alt key it also gets stuck. I need to press twice to get back to the canvas cursor. I'm not sure if it's a normal function, could you try it?

Thanks.

**Comment by pubpub-zz** [comment]
@chrisj0192 
Can you confirm the problem with the latest beta:
[ppInk64_LSEna16a.zip](https://github.com/pubpub-zz/ppInk/files/6828870/ppInk64_LSEna16a.zip)

If yes, can you make a video pls
